design-docs/configuration-dsl.md

```diff
@@ -0,0 +1,376 @@
+# Tool Chains
+> "A tool chain is a set of programming tools that are used to perform a complex software development task or to create a software product, which is typically another computer program or a set of related programs." - Wikipedia
+
+Tool chains consist of a compiler, linker, and libraries. It is possible to mix the preprocessor, compiler and linker of the same toolchain family
+(GCC, MSVC and Clang) to produce binaries for the current platform. It's also possible to mix the preprocessor, compiler and
+linker between tool chain family (Clang/C2, NASM) to produce binaries for the current platform. When targeting more exotic platform, you can use
+preprocessor, compiler, and linker of toolchains for those platforms. It's called cross-compiling. The compiler runs on a host machine targeting an
+entirely different architecture and/or operating system. Cross-compilation is all about the linker and sometimes the compiler. It means you could
+mix and match preprocessor with other compiler and linker.
+
+Clang family also bring another twist as you can dump the intermediate language (IR), execute transformations on it and feed it back to the backend
+of the compiler to complete the compilation. MSVC is moving in this direction as well, and GCC also supports this workflow.
+
+The proposed solution here would be to allow users to access the tool chain resolver more directly as well as enabling them to declare custom tool
+chains. Tool chains can be located in the system PATH, at predefined locations, identified from the registry (Windows) or identified through COM
+object (Windows), fetched remotely, installed through unattended installers or package manager or manually configured.
+
+## Open Issues
+ - The source dialect & ABI is its own beast to support. It's suggested modeling it separately and use compiler version to match variant
+ - Should it be named `tool chain` (Gradle) or `toolchain` (Wikipedia)?
+ - Should the tool chain and the libraries contained within (prebuilt libraries) be two separate concepts? Meaning the toolchain resolver can be used by prebuilt dependencies but would be tied together.
```

Comment:
Yes, absolutely. There are 2 abstract concepts here:

1. Tools that do some work.
2. Libraries that are used as input to that work.

Some tools implicitly provide a certain library when run, some tools do not, some tools can but can also be configured to not do so.

The same is true of tool chains. A tool chain typically bundles certain libraries along with certain tools, and often these bundled libraries are implicitly provided by the tools of the tool chain.

However, it is possible to mix and match, so that a given tool can use libraries that are not bundled in the same tool chain, including libraries that are bundled in other tool chains.

When we build something, there are 2 separate resolution steps:

1. Locate a tool that can do the work.
2. Locate all of the required libraries, taking into account the libraries that are implicitly provided by the tool, so that we can provide any libraries that are not implicitly provided, and also so we can _disable_ any implicit libraries that shouldn't be used.

Each of these resolution steps might use a chain of resolvers, which are simply providers that know how to find  tools or libraries. These chains would be separate - one for locating tools and one for locating libraries. A tool can be considered a provider of libraries, as can a tool chain, so when looking for a particular library there would be a resolver that would look at the libraries provided by the selected tool and tool chain to see if the library is already implicitly provided.